contracts/services/astroport-lper/src/msg.rs

```diff
@@ -0,0 +1,279 @@
+}
+
+impl ServiceConfig {
+    pub fn new(
+        input_addr: String,
+        output_addr: String,
+        pool_addr: String,
+        lp_config: LiquidityProviderConfig,
+    ) -> Self {
+        ServiceConfig {
+            input_addr,
+            output_addr,
+            pool_addr,
+            lp_config,
+        }
+    }
+
+    fn do_validate(&self, api: &dyn cosmwasm_std::Api) -> Result<(Addr, Addr), ServiceError> {
+        let input_addr = api.addr_validate(&self.input_addr)?;
+        let output_addr = api.addr_validate(&self.output_addr)?;
+        api.addr_validate(&self.pool_addr)?;
+
+        Ok((input_addr, output_addr))
+    }
+}
+
+#[cw_serde]
+pub struct LiquidityProviderConfig {
+    /// Pool type, old Astroport pools use Cw20 lp tokens and new pools use native tokens, so we specify here what kind of token we are going to get.
+    /// We also provide the PairType structure of the right Astroport version that we are going to use for each scenario
+    pub pool_type: PoolType,
+    /// Denoms of both native assets we are going to provide liquidity for
+    pub asset_data: AssetData,
+    /// Slippage tolerance when providing liquidity
+    pub slippage_tolerance: Option<Decimal>,
+}
+
+#[cw_serde]
+pub enum PoolType {
+    NativeLpToken(astroport::factory::PairType),
+    Cw20LpToken(astroport_cw20_lp_token::factory::PairType),
+}
+
+#[cw_serde]
+pub struct AssetData {
+    /// Denom of the first asset
+    pub asset1: String,
+    /// Denom of the second asset
+    pub asset2: String,
+}
+
+#[cw_serde]
+/// Validated service configuration
+pub struct Config {
+    pub input_addr: Addr,
+    pub output_addr: Addr,
+    pub pool_addr: Addr,
+    pub lp_config: LiquidityProviderConfig,
+}
+
+impl ServiceConfigInterface<ServiceConfig> for ServiceConfig {
+    /// This function is used to see if 2 configs are different
+    fn is_diff(&self, other: &ServiceConfig) -> bool {
+        !self.eq(other)
+    }
+}
+
+impl ServiceConfigValidation<Config> for ServiceConfig {
+    #[cfg(not(target_arch = "wasm32"))]
+    fn pre_validate(&self, api: &dyn cosmwasm_std::Api) -> Result<(), ServiceError> {
+        self.do_validate(api)?;
+        Ok(())
+    }
+
+    fn validate(&self, deps: Deps) -> Result<Config, ServiceError> {
+        let input_addr = deps.api.addr_validate(&self.input_addr)?;
+        let output_addr = deps.api.addr_validate(&self.output_addr)?;
+        let pool_addr = deps.api.addr_validate(&self.pool_addr)?;
+
+        ensure_correct_pool(
+            self.pool_addr.to_string(),
+            &self.lp_config.pool_type,
+            &self.lp_config.asset_data,
+            &deps,
+        )?;
+
+        Ok(Config {
+            input_addr,
+            output_addr,
+            pool_addr,
+            lp_config: self.lp_config.clone(),
+        })
+    }
+}
+
+impl OptionalServiceConfig {
+    pub fn update_config(self, deps: &DepsMut, config: &mut Config) -> Result<(), ServiceError> {
```

Comment:
You don't actually save the config here.

Reply:
Right, this is correct. It just updates it, saving is done in another step

Reply:
You are not using `OptionalServiceConfig` anywhere, so not sure what is going on here.

Reply:
It's being used now after the refactor.